### ISSUE TITLE
bundler aborts when x64-mingw32 gem isn't compatible with given Ruby version

### ISSUE THREAD

**Issue by larskanis** [oneclick/rubyinstaller2#7]
Nokogiri provides a source gem (platform `ruby`) and a gem for the `x64-mingw32` platform. The latter one is restricted to ruby versions 2.1 to 2.3, but the source gem has no ruby version restriction.

```sh
C:\Users\lars\railstest>ruby --version
ruby 2.4.0p0 (2016-12-24 revision 57164) [x64-mingw32]

C:\Users\lars\railstest>bundle --version
Bundler version 1.13.7

C:\Users\lars\railstest>gem list nokogiri
*** LOCAL GEMS ***
nokogiri (1.7.0.1)

C:\Users\lars\railstest>gem list -r nokogiri
*** REMOTE GEMS ***
nokogiri (1.7.0.1 ruby java x64-mingw32 x86-mingw32, 1.6.1 x86-mswin32-60, 1.4.4.1 x86-mswin32)

C:\Users\lars\railstest>rm Gemfile.lock

C:\Users\lars\railstest>bundle
Fetching gem metadata from https://rubygems.org/..........
Fetching version metadata from https://rubygems.org/..
Fetching dependency metadata from https://rubygems.org/.
Resolving dependencies.............................................
nokogiri-1.7.0.1-x64-mingw32 requires ruby version < 2.4, which is incompatible
with the current version, ruby 2.4.0p0
```

I would expect, that bundler chooses the gem for platform 'ruby' as the alternative without this ruby version constraint, but it refuses install altogether.

The only workaround I could find so far, is to install all dependent gems by `gem install` and run `bundle install --local` to generate the `Gemfile.lock`.

**Comment by MSP-Greg** [comment]
While working with my version (6.2, etc) of RI, I found a similar issue (Nokogiri & libxml-ruby) when using ruby-trunk (2.5.0), so I used (downloaded) local *.gem copies for install.  Never checked to see where the issue was, or a solution (maybe adding something to .gemrc?).  After that, I used local for all extension gems that had issues.

FYI, I'm trying to get the current version of RI2 to work, but I had to make changes to the original version because I have my msys64 folder on my d drive...

**Comment by larskanis** [comment]
I reported this issue upstream: https://github.com/bundler/bundler/issues/5337

**Comment by JunichiIto** [comment]
Just FYI, this is my temporary solution.

I tried "rails new" with Rails 5.1 against "ruby 2.4.1p111 (2017-03-22 revision 58053) [x64-mingw32]", but got this error while bundle install:

```
Gem::InstallError: nokogiri requires Ruby version < 2.4, >= 2.1.0.
An error occurred while installing nokogiri (1.7.1), and Bundler cannot
continue.
Make sure that `gem install nokogiri -v '1.7.1'` succeeds before bundling.
```

I previously installed Nokogiri 1.6.6.4:

```
c:\dev\rails-5-1-final-sandbox>gem list nokogiri

*** LOCAL GEMS ***

nokogiri (1.7.1, 1.7.0.1, 1.6.6.4 ruby x64-mingw32)
```

So I edited Gemfile to use Nokogiri 1.6.6.4

```ruby
# Gemfile
gem 'nokogiri', '1.6.6.4'
```

Then, bundle install succeeded and Rails 5.1 worked fine.

**Comment by MSP-Greg** [comment]
@JunichiIto 

Please see [Wiki FAQ](https://github.com/oneclick/rubyinstaller2/wiki/FAQ)

I've had Nokogiri loaded as a gem for a long time, so I don't recall if the package mingw-w64-x86_64-libxml2 is required  for install or not.

Re Bundler, having also used Ruby trunk for a long time, it's common for there to be errors or incorrect config info for new Ruby versions.  I would highly recommend installing Nokogiri 1.7.1.

[Ruby, Rails, AWS, Nokogiri docs](https://msp-greg.github.io/)

**Comment by JunichiIto** [comment]
@MSP-Greg Thank you for letting me know 😃

**Comment by MSP-Greg** [comment]
@JunichiIto  You're welcome.

Until @larskanis spent the time to write RubyInstaller2, there was no readily available version of Ruby 2.4 for Windows; one had to build it.  It's still not really available, so I'd expect 'special fun' with gems until it's in more common use.  Or, keep that wiki link handy...

As I said, I'm using trunk (2.5). I was just just working on a PR for YARD, and I decided to just rename Gemfile and Gemfile.lock rather than all the 'can't find this old gem version' messages...

**Comment by larskanis** [comment]
Luckily this issue is solved since bundler-2.2 :tada:
